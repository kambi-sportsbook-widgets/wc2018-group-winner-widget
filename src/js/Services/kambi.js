import { getEventsByFilter, getEvent } from 'kambi-offering-api-module'

class KambiService {
  /**
   * Fetches groups for given tournament.
   * @param {string} filter Tournament's filter
   * @param {number} criterionId Tournament's criterion identifier
   * @returns {Promise.<object[]>}
   */
  static getGroups(filter, criterionId) {
    return getEventsByFilter(`${filter}/all/all/competitions`)
      .then(competitions => {
        if (competitions == null) {
          throw new Error(
            `No tournament data available for supplied filter: ${filter}/all/all/competitions`
          )
        }
        const groupEvents = []

        // filter out group events with at least one betOffer and add event request to groupEvents
        const filteredEvents = competitions.events
          .filter(event => {
            const isGroupEvent = event.event.englishName.match(/Group ([A-Z])/)
            const hasBetOffers = event.betOffers && event.betOffers.length > 0
            return isGroupEvent && hasBetOffers
          })
          .forEach(event => {
            groupEvents.push(getEvent(event.event.id))
          })

        return Promise.all(groupEvents)
      })
      .then(groupEvents => {
        // filter betOffers by criterionId and sort groups by name
        const filteredByCriterionId = groupEvents
          .map(groupEvent => {
            groupEvent.betOffers = groupEvent.betOffers.filter(betOffer => {
              return betOffer.criterion.id === criterionId
            })
            groupEvent.groupName = groupEvent.event.englishName.match(
              /Group ([A-Z])/
            )[1]

            if (groupEvent.betOffers.length < 2 || !groupEvent.groupName) {
              throw new Error('Could not find matching betoffers or group name')
              return
            }
            return groupEvent
          })
          .sort((a, b) => {
            return a.event.englishName.localeCompare(b.event.englishName)
          })

        return Promise.resolve(filteredByCriterionId)
      })
      .catch(err => {
        return Promise.reject(err)
      })
  }

  /**
   * Returns home team name of tournament's closest match.
   * @param {string} filter Tournament's filter
   * @returns {Promise.<string|null>}
   */
  static getNextMatches(filter) {
    return getEventsByFilter(`${filter}/all/all/matches`)
      .then(matches => {
        if (matches == null) {
          throw new Error(
            `No data available for supplied filter: ${filter}/all/all/matches`
          )
        }
        const currentTime = new Date(Date.now())

        return matches.events
          .filter(m => {
            const eventStart = new Date(m.event.start)
            return (
              m.event.tags.indexOf('MATCH') !== -1 &&
              eventStart &&
              eventStart > currentTime
            )
          })
          .sort((a, b) => new Date(a.event.start) - new Date(b.event.start))
      })
      .then(matches => {
        return matches.length > 0 ? matches : null
      })
  }
}

export default KambiService
