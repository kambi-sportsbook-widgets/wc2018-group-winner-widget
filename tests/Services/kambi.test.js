import kambi from '../../src/js/Services/kambi';
import { coreLibrary, offeringModule, widgetModule } from 'kambi-widget-core-library';


jest.mock('kambi-widget-core-library', () => ({
   coreLibrary: {
      config: {
         routeRoot: ''
      }
   },
   offeringModule: {
      getHighlight: jest.fn(),
      getEventsByFilter: jest.fn()
   },
   widgetModule: {
      createFilterUrl: jest.fn()
   }
}));

const mockHighlightGroup = {pathTermId: 'test'};

const mockEvent = {
   event: {
      type: 'ET_MATCH'
   }
};


describe('Kambi get group services', () => {

   beforeEach(() => {
      offeringModule.getEventsByFilter = jest.fn();
   });

   it('returns events correctly', () => {
      const filter = 'testTournament';

      offeringModule.getEventsByFilter.mockReturnValueOnce(new Promise(resolve => resolve({events: []})));
      offeringModule.getEventsByFilter.mockReturnValueOnce(new Promise(resolve => resolve({events: [
         mockEvent,
         Object.assign({}, mockEvent, {event: {type: 'ET_COMPETITION'}})
      ]})));

      expect(offeringModule.getEventsByFilter).not.toHaveBeenCalled();

      return kambi.getGroups(filter)
         .then((events) => {
            expect(events).toMatchSnapshot();
            expect(offeringModule.getEventsByFilter).toHaveBeenCalledTimes(1);
            expect(offeringModule.getEventsByFilter).toHaveBeenLastCalledWith('testTournament/all/all/competitions');
         });
   });

   it('behaves correctly when no events were found', () => {
      const filter = 'testTournament';

      offeringModule.getEventsByFilter.mockReturnValueOnce(new Promise(resolve => resolve({events: []})));

      expect(offeringModule.getEventsByFilter).not.toHaveBeenCalled();

      return kambi.getGroups(filter)
         .then((events) => {
            throw new Error('Not supposed to reach here');
         })
         .catch((err) => {
            expect(err.message === `No tournament data available for supplied filter: ${filter}/all/all/competitions`);
         })
   });

});


describe('Kambi getNextMatches', () => {

   beforeEach(() => {
      offeringModule.getEventsByFilter = jest.fn();
   });

   it('returns events correctly', () => {
      const filter = 'testTournament';

      offeringModule.getEventsByFilter.mockReturnValueOnce(new Promise(resolve => resolve({events: []})));
      offeringModule.getEventsByFilter.mockReturnValueOnce(new Promise(resolve => resolve({events: [
         mockEvent,
         Object.assign({}, mockEvent, {event: {type: 'ET_COMPETITION'}})
      ]})));

      expect(offeringModule.getEventsByFilter).not.toHaveBeenCalled();

      return kambi.getNextMatches(filter)
         .then((events) => {
            expect(events).toMatchSnapshot();
            expect(offeringModule.getEventsByFilter).toHaveBeenCalledTimes(1);
            expect(offeringModule.getEventsByFilter).toHaveBeenLastCalledWith('testTournament/all/all/matches');
         });
   });

   it('behaves correctly when no events were found', () => {
      const filter = 'testTournament';

      offeringModule.getEventsByFilter.mockReturnValueOnce(new Promise(resolve => resolve({events: []})));

      expect(offeringModule.getEventsByFilter).not.toHaveBeenCalled();

      return kambi.getNextMatches(filter)
         .then((events) => {
            throw new Error('Not supposed to reach here');
         })
         .catch((err) => {
            expect(err.message === `No data available for supplied filter: ${filter}/all/all/matches`);
         })
   });

});