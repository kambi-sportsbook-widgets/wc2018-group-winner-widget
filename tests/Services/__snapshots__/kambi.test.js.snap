// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kambi get group services returns events correctly 1`] = `Array []`;

exports[`Kambi getNextMatches returns events correctly 1`] = `null`;
